langwatch/src/pages/settings/components/PermissionSelector.tsx

```diff
@@ -0,0 +1,174 @@
+import {
+  Box,
+  Fieldset,
+  HStack,
+  Separator,
+  Text,
+  VStack,
+} from "@chakra-ui/react";
+import { Info } from "react-feather";
+import { Checkbox } from "../../../components/ui/checkbox";
+import { Tooltip } from "../../../components/ui/tooltip";
+import type { Permission, Resource, Action } from "../../../server/api/rbac";
+import {
+  orderedResources,
+  getValidActionsForResource,
+} from "../../../utils/permissionsConfig";
+
+/**
+ * PermissionSelector component
+ *
+ * Single Responsibility: Provides an interactive interface for selecting and managing permissions
+ */
+export function PermissionSelector({
+  selectedPermissions,
+  onChange,
+}: {
+  selectedPermissions: Permission[];
+  onChange: (permissions: Permission[]) => void;
+}) {
+  // Helper function to safely create permission strings
+  const createPermission = (resource: Resource, action: Action): Permission => {
+    return `${resource}:${action}`;
+  };
+  const groupedPermissions: Record<Resource, Permission[]> = {} as Record<
+    Resource,
+    Permission[]
+  >;
+
+  // Use orderedResources from shared config (PLAYGROUND hidden, ORG/TEAM omitted)
+  const resourceOrder: Resource[] = orderedResources;
+
+  // Group permissions by resource using the correct valid actions
+  resourceOrder.forEach((resource) => {
+    const validActions = getValidActionsForResource(resource);
+    groupedPermissions[resource] = validActions.map((action) =>
+      createPermission(resource, action),
+    );
+  });
+
+  const togglePermission = (permission: Permission) => {
```

Comment:
This function is a bit tough to read easily, and seems like it does way more than simply toggle. Consider better naming/documentation/splitting